core/src/peer_id.rs

```diff
@@ -77,57 +72,35 @@ impl PeerId {
-        // `PeerId`s of different hashes equal, which makes it possible to connect through
-        // secio or noise to nodes with an identity hash. Starting from version 0.17, rust-libp2p
-        // will switch to not hashing the key (i.e. the correct behaviour).
-        // In other words, rust-libp2p 0.16 is compatible with all versions of rust-libp2p.
-        // Rust-libp2p 0.12 and below is **NOT** compatible with rust-libp2p 0.17 and above.
-        let (hash_algorithm, canonical_algorithm) = if key_enc.len() <= MAX_INLINE_KEY_LENGTH {
-            (Code::Identity, Some(Code::Sha2_256))
+        let hash_algorithm = if key_enc.len() <= MAX_INLINE_KEY_LENGTH {
+            Code::Identity
         } else {
-            (Code::Sha2_256, None)
+            Code::Sha2_256
         };
 
-        let canonical = canonical_algorithm.map(|alg|
-            alg.digest(&key_enc));
-
         let multihash = hash_algorithm.digest(&key_enc);
 
-        PeerId { multihash, canonical }
+        PeerId { multihash }
     }
 
     /// Checks whether `data` is a valid `PeerId`. If so, returns the `PeerId`. If not, returns
     /// back the data as an error.
     pub fn from_bytes(data: Vec<u8>) -> Result<PeerId, Vec<u8>> {
-        match multihash::Multihash::from_bytes(data) {
-            Ok(multihash) => {
-                if multihash.algorithm() == multihash::Code::Sha2_256 {
-                    Ok(PeerId { multihash, canonical: None })
-                }
-                else if multihash.algorithm() == multihash::Code::Identity {
-                    let canonical = Sha2_256::digest(&multihash.digest());
-                    Ok(PeerId { multihash, canonical: Some(canonical) })
-                } else {
-                    Err(multihash.into_bytes())
-                }
-            }
+        match Multihash::from_bytes(data) {
+            Ok(multihash) => match multihash.algorithm() {
+                Code::Sha2_256 | Code::Identity => Ok(PeerId { multihash }),
+                _ => Err(multihash.into_bytes())
+            },
             Err(err) => Err(err.data),
         }
     }
 
     /// Turns a `Multihash` into a `PeerId`. If the multihash doesn't use the correct algorithm,
     /// returns back the data as an error.
-    pub fn from_multihash(data: multihash::Multihash) -> Result<PeerId, multihash::Multihash> {
-        if data.algorithm() == multihash::Code::Sha2_256 {
-            Ok(PeerId { multihash: data, canonical: None })
-        } else if data.algorithm() == multihash::Code::Identity {
-            let canonical = Sha2_256::digest(data.digest());
-            Ok(PeerId { multihash: data, canonical: Some(canonical) })
-        } else {
-            Err(data)
+    pub fn from_multihash(data: Multihash) -> Result<PeerId, Multihash> {
+        match data.algorithm() {
+            Code::Sha2_256 | Code::Identity => Ok(PeerId { multihash: data }),
```

Comment:
This isn't the case right now, but maybe we should check that `data.len() <= MAX_INLINE_KEY_LENGTH` when the algorithm is `identity`?

Reply:
Sounds good to me.

Reply:
See https://github.com/libp2p/rust-libp2p/pull/1608/commits/c251d8f730fdec90fb82d53ecb274fd474270936.